gufe/component.py

```diff
@@ -4,37 +4,103 @@
 import abc
 from typing import Union
 
+from .base import GufeTokenizableMixin
 
-class Component(abc.ABC):
+
+COMPONENT_REGISTRY = {}
+
+def register_component(cls):
+    COMPONENT_REGISTRY[cls.__name__] = cls
+
+    
+class _ComponentMeta(type):
+    def __new__(cls, name, bases, classdict):
+        componentcls = super().__new__(cls, name, bases, classdict)
+        register_component(componentcls)
+        return componentcls
+
+
+class _ABCComponentMeta(_ComponentMeta, abc.ABCMeta):
+    # required to make use of abc.ABC in classes that use _ComponentMeta metaclass
+    # see https://stackoverflow.com/questions/57349105/python-abc-inheritance-with-specified-metaclass 
+    ...
+        
+
+class Component(abc.ABC, GufeTokenizableMixin, metaclass=_ABCComponentMeta):
     """Base class for members of a ChemicalSystem"""
 
+    def __init__(self):
+
+        # since __hash__ isn't inheritable, monkey-patching is the best we can
+        # do besides defining it in every subclass
+        self.__class__.__hash__ = Component.__hash__
+
     def __repr__(self):
         return f"{self.__class__.__name__}(name={self.name})"
 
-    @abc.abstractmethod
-    def __hash__(self):
-        pass
-
     def __lt__(self, other):
         return hash(self) < hash(other)
 
-    @abc.abstractmethod
     def __eq__(self, other):
-        pass
+        if not isinstance(other, self.__class__):
+            return NotImplemented("Component comparisons must be of the same type")
 
-    @abc.abstractmethod
-    def to_dict(self) -> dict:
-        pass
+        d = self.to_dict()
+        do = other.to_dict()
+
+        if set(d.keys()) == set(do.keys()):
+            for key in d:
+                if d[key] != do[key]:
+                    return False
+
+            return True
```

Comment:
if keys don't match this falls through, does that default to False?  just do `return d == do`?

Reply:
Ah good point. Fixed!